src/Features/Core/Portable/RemoveUnusedParametersAndValues/AbstractRemoveUnusedParametersAndValuesDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,856 @@
+                         (invocation.TargetMethod.ContainingType.OriginalDefinition == _symbolStartAnalyzer._interlockedType ||
+                          invocation.TargetMethod.ContainingType.OriginalDefinition == _symbolStartAnalyzer._immutableInterlockedType)))
+                    {
+                        return;
+                    }
+
+                    // "Expression value is never used"
+                    var properties = s_propertiesMap[(_options.UnusedValueExpressionStatementPreference, isUnusedLocalAssignment: false, isRemovableAssignment: false)];
+                    var diagnostic = DiagnosticHelper.Create(s_expressionValueIsUnusedRule,
+                                                             value.Syntax.GetLocation(),
+                                                             _options.UnusedValueExpressionStatementSeverity,
+                                                             additionalLocations: null,
+                                                             properties);
+                    context.ReportDiagnostic(diagnostic);
+                }
+
+                private void AnalyzeDelegateCreationOrAnonymousFunction(OperationAnalysisContext operationAnalysisContext)
+                    => _hasDelegateCreationOrAnonymousFunction = true;
+
+                private void AnalyzeConversion(OperationAnalysisContext operationAnalysisContext)
+                {
+                    if (_hasConversionFromDelegateTypeToNonDelegateType)
+                    {
+                        return;
+                    }
+
+                    var conversion = (IConversionOperation)operationAnalysisContext.Operation;
+                    if (conversion.Operand.Type.IsDelegateType() &&
+                        !conversion.Type.IsDelegateType())
+                    {
+                        _hasConversionFromDelegateTypeToNonDelegateType = true;
+                    }
+                }
+
+                private void AnalyzeFieldOrPropertyReference(OperationAnalysisContext operationAnalysisContextContext)
+                {
+                    var fieldOrPropertyReference = operationAnalysisContextContext.Operation;
+                    if (fieldOrPropertyReference.Type.IsDelegateType() &&
+                        fieldOrPropertyReference.Parent is ISimpleAssignmentOperation simpleAssignment &&
+                        simpleAssignment.Target == fieldOrPropertyReference)
+                    {
+                        _escapedDelegate = true;
+                    }
+                }
+                private void AnalyzeOperationBlockEnd(OperationBlockAnalysisContext context)
+                {
+                    var isComputingUnusedParams = _options.IsComputingUnusedParams(context.OwningSymbol);
+                    if (_options.UnusedValueExpressionStatementPreference == UnusedValuePreference.None &&
+                        !isComputingUnusedParams)
+                    {
+                        return;
```

Comment:
this whole check seems so bizarre.  how would we even get here if the options weren't set to make us register these callbacks?

Reply:
Couple of things: there is a typo here, I actually meant `UnusedValueAssignmentPreference`, which is fixed now. Also, regarding your question, this is the first place where we check the value for this option. We do not check it at operation block start as that registers options for different callbacks that can report 2 diagnostics: unused value from expression statement and unused value assignments. Former is reported in `AnalyzeExpressionStatement` and the latter reported here, and each of these callbacks have the option preference as first check.

---
In reply to: [230605502](https://github.com/dotnet/roslyn/pull/30777#discussion_r230605502) [](ancestors = 230605502)

Reply:
great thanks.  sounds like that would have been a test hole as well.  can you fill it?  thanks!

Reply:
This is just an optimization that checks all of multiple options are disabled (are both unused params and unused values set to disabled?). Each method reporting individual diagnostic then checks if individual option for diagnostic it is reporting is disabled - so end user experience is not changed by keeping or removing this upfront check.

---
In reply to: [230987892](https://github.com/dotnet/roslyn/pull/30777#discussion_r230987892) [](ancestors = 230987892)